lptree.pas

```diff
@@ -5652,11 +5763,35 @@ function TLapeTree_For.CompileBody(var Offset: Integer): TResVar;
   IncDec: TLapeTree_InternalMethod;
   tmpBody: TLapeTree_Base;
   tmpContinueStatements: TLapeFlowStatementList;
+  limVar,tmpVar: TResVar;
 begin
   Assert((FCondition <> nil) and (FCondition is TLapeTree_Operator));
   Assert((TLapeTree_Operator(FCondition).Right <> nil) and (TLapeTree_Operator(FCondition).Right is TLapeTree_ResVar));
 
   FStartBodyOffset := FCompiler.Emitter.CheckOffset(Offset);
+  
+  if walkIn then
+  begin
+    FLimit.CompileToTempVar(Offset, limVar);
+    FCounter.CompileToTempVar(Offset, tmpVar);
+
+    with TLapeTree_Operator.Create(op_Assign, FCompiler, @_DocPos) do
+    try
+      Left := TLapeTree_ResVar.Create(tmpVar.IncLock(), FCounter);
+      Right := TLapeTree_Operator.Create(op_Index, FCompiler);
+      with TLapeTree_Operator(Right) do
+      begin
+        Left := TLapeTree_ResVar.Create(limVar.IncLock(), FLimit);
+        Right := TLapeTree_ResVar.Create(TLapeTree_Operator(FCondition).Left.Compile(Offset).IncLock(), FCondition);
+      end;
+      Compile(Offset);
+    finally
+      Free();
+      limVar.Spill(1);
+      tmpVar.Spill(1);
```

Comment:
This spill is too early. `tmpVar` needs to be valid throughout the whole loop body, and can only be spilled afterwards.